api/v1alpha1/utils.go

```diff
@@ -42,3 +46,20 @@ func BoolToString(b *bool) string {
 
 	return "false"
 }
+
+// IsEqualStruct is a util fonction that returns whether 2 structures are the same
+// We compare the marchaled results to avoid traversing all fields and be agnostic of the struct.
+func IsEqualStruct(in interface{}, cmp interface{}) bool {
```

Comment:
Looking at the PR, it looks like `IsEqualStruct` is only used to compare `userInput` and `emptyStruct`, like `v1alpha1.IsEqualStruct(dd.Spec.Agent, v1alpha1.DatadogAgentSpecAgentSpec{})`.
Doing JSON Marshalling is costly in terms of CPU and I wonder if we could do better here.

Reply:
I agree, it's not ideal... Because of the intricacy of the whole structure (with pointers of other structures etc), I decided to go the marchalling route to make sure that I would compare apple to apple. 
I did look into a few alternatives with build tags, interfaces etc but that was the best I could find.
I did isolate the function on purpose though, in case we figure out something better.

Reply:
I would expect https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go to work in this scenario. We could extend the `Semantic` var with more types if not.

Reply:
I looked into it a bit - It would probably work indeed. Though as it uses reflects under the hood I would be curious to benchmark the change. I will make a card so we can evaluate changing this bit of logic.